File: Submissions/809_Expressive_Words.cpp
class Solution {
public:
    int expressiveWords(string S, vector<string>& words) {
        int result = 0;
        int S_length = S.size();
        for(auto word : words){
            int word_length = word.size();
            if(S_length < word_length) continue;
            int i=0, j=0;
            for(; i<S.size(); ++i){
                if(S[i] == word[j]) ++j;
                else if(i>0 && S[i]==S[i-1] && i+1<S_length && S[i]==S[i+1]) ++i;
                else if(i>1 && S[i]==S[i-1] && S[i]==S[i-2]) continue;
                else break;
            }
            
            if(i==S_length && j==word_length) ++result;
        }
        return result;
    }
};

File: Kotlin/809_Expressive_Words.kt
class Solution {
    fun expressiveWords(S: String, words: Array<String>): Int {
        var result = 0
        val SLength = S.length
        for(word in words){
            val wordLength = word.length
            if(SLength < wordLength) continue
            var i = 0
            var j = 0
            while(i<SLength){
                if(j<wordLength && S[i] == word[j]) {
                    ++j
                    ++i
                }
                else if(i>0 && S[i]==S[i-1] && i+1<SLength && S[i]==S[i+1]) i+=2
                else if(i>1 && S[i]==S[i-1] && S[i]==S[i-2]) ++i
                else break
            }
            
            if(i==SLength && j == wordLength) ++result
        }
        
        return result
    }
}